[MSE][Mac] Add TextTrack support to SourceBufferPrivateAVFObjC.

eric-carlson opened the pull request:
#### 6d6af3f630e58341c98f3fa29ac21657c1911961
<pre>
[MSE][Mac] Add TextTrack support to SourceBufferPrivateAVFObjC.
<a href="https://bugs.webkit.org/show_bug.cgi?id=125161">https://bugs.webkit.org/show_bug.cgi?id=125161</a>
<a href="https://rdar.apple.com/137769578">rdar://137769578</a>

Reviewed by Jean-Yves Avenard.

Add general TextTrack support to SourceBufferPrivateAVFObjC, and add support for muxed
WebVTT tracks.

* LayoutTests/media/media-source/content/bip-bop-webvtt-frag.mp4: Added.
* LayoutTests/media/media-source/media-managedmse-webvtt-track-expected.txt: Added.
* LayoutTests/media/media-source/media-managedmse-webvtt-track.html: Added.
* LayoutTests/platform/glib/TestExpectations: Skip new test.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Add new runtime setting.

* Source/WTF/wtf/Logger.h:
(WTF::LogArgument::toString): Support logging a StringView.

* Source/WebCore/Headers.cmake:

* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::createSourceBufferPrivate):

* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm: Add AVMediaTypeText.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/html/track/InbandTextTrack.cpp:
(WebCore::InbandTextTrack::create): Add `CueFormat::Unknown` to switch statement.

* Source/WebCore/html/track/InbandWebVTTTextTrack.cpp:
(WebCore::InbandWebVTTTextTrack::parseWebVTTFileHeader): Added.
* Source/WebCore/html/track/InbandWebVTTTextTrack.h:

* Source/WebCore/platform/graphics/InbandTextTrackPrivate.h: Add `CueFormat::Unknown`
so `m_format` can be initialized.

* Source/WebCore/platform/graphics/InbandTextTrackPrivate.serialization.in: Serialize
`CueFormat::Unknown`.

* Source/WebCore/platform/graphics/MediaSourceConfiguration.h:
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
* Source/WebCore/platform/graphics/TrackPrivateBase.cpp:
(WebCore::TrackPrivateBase::notifyMainThreadClient):

* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.h:
* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.mm:
(WebCore::AVTrackPrivateAVFObjCImpl::textKindForAVAssetTrack): Determine if a AVAssetTrack
has captions or subtitles based on its characteristics.
(WebCore::AVTrackPrivateAVFObjCImpl::textKindForAVMediaSelectionOption): Ditto, for an
AVMediaSelectionOption.
(WebCore::AVTrackPrivateAVFObjCImpl::textKind const): Return TextTrack.kind.

* Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
(WebCore::InbandTextTrackPrivateAVF::InbandTextTrackPrivateAVF): Make the AVFInbandTrackParent
parameter last and give it a default value of nullptr.
(WebCore::InbandTextTrackPrivateAVF::processCue): processNativeSamples -&gt; processVTTSamples.
(WebCore::InbandTextTrackPrivateAVF::setMode): Always call through to the base class.
(WebCore::InbandTextTrackPrivateAVF::processVTTFileHeader): Split out of processNativeSamples.
(WebCore::InbandTextTrackPrivateAVF::processVTTSample): Ditto.
(WebCore::InbandTextTrackPrivateAVF::processVTTSamples): Renamed from processNativeSamples.
(WebCore::InbandTextTrackPrivateAVF::readVTTSampleBuffer): Renamed from readNativeSampleBuffer.
(WebCore::InbandTextTrackPrivateAVF::processNativeSamples): Deleted.
(WebCore::InbandTextTrackPrivateAVF::readNativeSampleBuffer): Deleted.
* Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.h:
(WebCore::InbandTextTrackPrivateAVF::setKind):
(WebCore::InbandTextTrackPrivateAVF::setId):
(WebCore::InbandTextTrackPrivateAVF::setLabel):
(WebCore::InbandTextTrackPrivateAVF::setLanguage):
(isType):

* Source/WebCore/platform/graphics/avfoundation/VideoTrackPrivateAVF.h: Remove an
unnecessary blank line.

* Source/WebCore/platform/graphics/avfoundation/objc/InbandTextTrackPrivateAVFObjC.h:
(WebCore::InbandTextTrackPrivateAVFObjC::create): Add version that takes an AVAssetTrack.
* Source/WebCore/platform/graphics/avfoundation/objc/InbandTextTrackPrivateAVFObjC.mm:
(WebCore::InbandTextTrackPrivateAVFObjC::InbandTextTrackPrivateAVFObjC): Switch parameter
order.
(WebCore::InbandTextTrackPrivateAVFObjC::disconnect): Call base class first, clear `m_assetTrack`.
(WebCore::InbandTextTrackPrivateAVFObjC::mediaType const):
(WebCore::InbandTextTrackPrivateAVFObjC::hasMediaCharacteristic const):
(WebCore::InbandTextTrackPrivateAVFObjC::kind const):
(WebCore::InbandTextTrackPrivateAVFObjC::isClosedCaptions const):
(WebCore::InbandTextTrackPrivateAVFObjC::isSDH const): Use hasMediaCharacteristic.
(WebCore::InbandTextTrackPrivateAVFObjC::containsOnlyForcedSubtitles const): Ditto.
(WebCore::InbandTextTrackPrivateAVFObjC::isMainProgramContent const): Ditto.
(WebCore::InbandTextTrackPrivateAVFObjC::isEasyToRead const): Ditto.
(WebCore::InbandTextTrackPrivateAVFObjC::label const): Simplify logic.
(WebCore::InbandTextTrackPrivateAVFObjC::language const): Use static methods in AVTrackPrivateAVFObjCImpl.
(WebCore::InbandTextTrackPrivateAVFObjC::isDefault const): Null-check m_mediaSelectionGroup
and m_mediaSelectionOption.

* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:
(WebCore::MediaSourcePrivateAVFObjC::addSourceBuffer):

* Source/WebCore/platform/graphics/avfoundation/objc/OutOfBandTextTrackPrivateAVF.h:
(WebCore::OutOfBandTextTrackPrivateAVF::OutOfBandTextTrackPrivateAVF):

* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferParserAVFObjC.mm:
(WebCore::MediaDescriptionAVFObjC::MediaDescriptionAVFObjC): Use -[AVAssetTrack mediaType]
instead of -[AVAssetTrack hasMediaCharacteristic] to get media type.
(WebCore::SourceBufferParserAVFObjC::SourceBufferParserAVFObjC): Add configuration parameter.
(WebCore::SourceBufferParserAVFObjC::didParseStreamDataAsAsset): Add support for WebVTT
tracks. Always log when we ignore an unknown track type.

* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::setTrackChangeCallbacks): Add a comment about why
we don&apos;t do anything when a text track mode changes.
(WebCore::SourceBufferPrivateAVFObjC::precheckInitializationSegment): Support text tracks.
(WebCore::SourceBufferPrivateAVFObjC::isMediaSampleAllowed const): Deal with text samples.
(WebCore::SourceBufferPrivateAVFObjC::clearTracks): Support text tracks.
(WebCore::SourceBufferPrivateAVFObjC::trackDidChangeEnabled): Fix indentation.
(WebCore::SourceBufferPrivateAVFObjC::isTextTrack const):

* Source/WebCore/platform/graphics/avfoundation/objc/TextTrackPrivateMediaSourceAVFObjC.cpp: Added.
(WebCore::TextTrackPrivateMediaSourceAVFObjC::TextTrackPrivateMediaSourceAVFObjC):
(WebCore::TextTrackPrivateMediaSourceAVFObjC::resetPropertiesFromTrack):
* Source/WebCore/platform/graphics/avfoundation/objc/TextTrackPrivateMediaSourceAVFObjC.h:

* Source/WebCore/platform/graphics/cocoa/SourceBufferParser.cpp:
(WebCore::SourceBufferParser::create): Add configuration parameter.
* Source/WebCore/platform/graphics/cocoa/SourceBufferParser.h:

* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.cpp:
(WebCore::MediaSourcePrivateGStreamer::addSourceBuffer): Ditto
* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.h:

* Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.cpp:
(WebCore::MockMediaSourcePrivate::addSourceBuffer): Ditto.
* Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.h:

* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::loadMediaSource): Remove webMParserEnabled parameter.
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:

* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:
(WebKit::RemoteMediaSourceProxy::RemoteMediaSourceProxy):
(WebKit::RemoteMediaSourceProxy::addSourceBuffer): Add configuration parameter.
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.messages.in:

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in: Add MediaSourceConfiguration.

* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::load): Add configuration parameter.

* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
(WebKit::MediaSourcePrivateRemote::addSourceBuffer): Ditto.
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.h:

Canonical link: <a href="https://commits.webkit.org/290705@main">https://commits.webkit.org/290705@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e16767451158fadfaff6ad4d73f3d37d56a6099d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90719 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10256 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45665 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95734 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41505 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92772 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10650 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18573 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69798 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27342 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93720 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8126 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82262 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50140 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7897 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36647 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40634 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/83536 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78197 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37711 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97564 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/89499 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17914 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13153 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78827 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18173 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78092 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78024 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19289 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22468 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/21093 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17923 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23266 "Found 2 new failures in platform/graphics/avfoundation/objc/SourceBufferParserAVFObjC.mm") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/111988 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17662 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/32601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21118 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19446 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
```diff
@@ -488,19 +488,89 @@ void InbandTextTrackPrivateAVF::resetCueValues()
+
+bool InbandTextTrackPrivateAVF::processVTTFileHeader(CMFormatDescriptionRef formatDescription)
+{
+    using namespace PAL;
```
probably not needed as the only call to CM method is already prefixed by `PAL::`

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
```diff
@@ -488,19 +488,89 @@ void InbandTextTrackPrivateAVF::resetCueValues()
+    if (!extensions)
+        return false;
+
+    CFDictionaryRef sampleDescriptionExtensions = static_cast<CFDictionaryRef>(CFDictionaryGetValue(extensions, kCMFormatDescriptionExtension_SampleDescriptionExtensionAtoms));
```
```suggestion
    auto sampleDescriptionExtensions = static_cast<CFDictionaryRef>(CFDictionaryGetValue(extensions, kCMFormatDescriptionExtension_SampleDescriptionExtensionAtoms));
```

eric-carlson replied:
Oh yeah, there are several places where I moved old code around that can now use `auto`. Will update.

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
```diff
@@ -515,92 +585,25 @@ void InbandTextTrackPrivateAVF::processNativeSamples(CFArrayRef nativeSamples, c
-
-            m_sampleInputBuffer.remove(0, (size_t)boxLength);
-        }
+        CMSampleBufferRef sampleBuffer = reinterpret_cast<CMSampleBufferRef>(const_cast<void*>(CFArrayGetValueAtIndex(nativeSamples, i)));
```
```suggestion
        auto sampleBuffer = reinterpret_cast<CMSampleBufferRef>(const_cast<void*>(CFArrayGetValueAtIndex(
```

cdumez replied:
Ideally `RetainPtr`?

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
```diff
@@ -611,9 +614,8 @@ bool InbandTextTrackPrivateAVF::readNativeSampleBuffer(CFArrayRef nativeSamples,
-    buffer = ArrayBuffer::create(m_sampleInputBuffer);
-
     formatDescription = CMSampleBufferGetFormatDescription(sampleBuffer);
+    duration = PAL::toMediaTime(timingInfo.duration);
```
Should we use the duration of the entire CMSampleBuffer rather than the first sub-sample one ? 
Should we assert that this CMSampleBuffer only contains a single sample?

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
```diff
@@ -541,6 +558,13 @@ static inline bool shouldAddContentKeyRecipients()
+
+    for (auto& pair : m_textTracks) {
+        RefPtr track = pair.second;
+        if (auto player = this->player())
```
```suggestion
        if (RefPtr player = this->player())
```

to align with new code style

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/objc/TextTrackPrivateMediaSourceAVFObjC.h:
```diff
@@ -0,0 +1,63 @@
+        return adoptRef(*new TextTrackPrivateMediaSourceAVFObjC(track, format));
+    }
+
+    virtual ~TextTrackPrivateMediaSourceAVFObjC();
```
`virtual` is redundant it as base class has (it must)

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/objc/TextTrackPrivateMediaSourceAVFObjC.h:
```diff
@@ -0,0 +1,63 @@
+class SourceBufferPrivateAVFObjC;
+
+class TextTrackPrivateMediaSourceAVFObjC final : public InbandTextTrackPrivateAVF {
+    WTF_MAKE_TZONE_ALLOCATED(TextTrackPrivateMediaSourceAVFObjC);
```
I don't believe we need this as the base class is already doing it

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
```diff
@@ -488,19 +488,89 @@ void InbandTextTrackPrivateAVF::resetCueValues()
+    if (!sampleDescriptionExtensions)
+        return false;
+
+    CFDataRef webvttHeaderData = static_cast<CFDataRef>(CFDictionaryGetValue(sampleDescriptionExtensions, CFSTR("vttC")));
```
```suggestion
    auto webvttHeaderData = static_cast<CFDataRef>(CFDictionaryGetValue(sampleDescriptionExtensions, CFSTR("vttC")));
```

cdumez replied:
Ideally we'd even use `RetainPtr` I think.

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
```diff
@@ -488,19 +488,89 @@ void InbandTextTrackPrivateAVF::resetCueValues()
-    m_owner->trackModeChanged();
+    while (true) {
+        RefPtr<ArrayBuffer> buffer = ArrayBuffer::create(m_sampleInputBuffer);
+        auto view = JSC::DataView::create(WTFMove(buffer), 0, buffer->byteLength());
```
```suggestion
        Ref view = JSC::DataView::create(WTFMove(buffer), 0, buffer->byteLength());
```

(new guidelines)

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
```diff
@@ -488,19 +488,89 @@ void InbandTextTrackPrivateAVF::resetCueValues()
+        if (!peekResult)
+            break;
+
+        auto type = peekResult.value().first;
```
```suggestion
        auto type = peekResult->first;
```

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.h:
```diff
@@ -97,28 +107,38 @@ class InbandTextTrackPrivateAVF : public InbandTextTrackPrivate {
 
     Vector<Ref<InbandGenericCue>> m_cues;
-    AVFInbandTrackParent* m_owner;
+    AVFInbandTrackParent* m_owner { nullptr };
```
Should consider making `AVFInbandTrackParent` CanMakeWeakPtr to align with the new safety style

cdumez replied:
Yes, or `CanMakeCheckedPtr` if you don't need the automatic nulling out on destruction.

jyavenard reviewed (approved):
LGTM.

We need extra tests for MSE in a worker.

cdumez commented on Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
```diff
@@ -488,19 +488,89 @@ void InbandTextTrackPrivateAVF::resetCueValues()
+    auto header = makeString(std::span { CFDataGetBytePtr(webvttHeaderData), length }, "\n\n"_s);
+
+    INFO_LOG(LOGIDENTIFIER, "VTT header ", header);
+    notifyMainThreadClient([&](auto& client) {
```
I'm not familiar with `notifyMainThreadClient()` but I assume it dispatches an asynchronous task (maybe even to another thread). Either way, if so, capturing header by reference doesn't seem safe and we'd likely want to capture `header = WTFMove(header)` (if no multi-threading) or `header = crossThreadCopy(WTFMove(header))` (if multithreaded).

eric-carlson replied:
Excellent point. I will fix this, the one you mention below, and the other places in this file that don't show up in this PR because I didn't modify them.

cdumez commented on Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
```diff
@@ -488,19 +488,89 @@ void InbandTextTrackPrivateAVF::resetCueValues()
 
-    m_owner->trackModeChanged();
+    while (true) {
+        RefPtr<ArrayBuffer> buffer = ArrayBuffer::create(m_sampleInputBuffer);
```
Can likely omit `<ArrayBuffer>`

cdumez commented on Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
```diff
@@ -488,19 +488,89 @@ void InbandTextTrackPrivateAVF::resetCueValues()
+        if (type == ISOWebVTTCue::boxTypeName()) {
+            ISOWebVTTCue cueData = ISOWebVTTCue(presentationTime, duration);
+            cueData.read(view);
+            notifyMainThreadClient([&](auto& client) {
```
Same concern here about capturing by reference in notifyMainThreadClient().

cdumez commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
```diff
@@ -945,6 +969,11 @@ static inline bool shouldAddContentKeyRecipients()
 
+bool SourceBufferPrivateAVFObjC::isTextTrack(TrackID trackID) const
+{
+    return m_textTracks.find(trackID) != m_textTracks.end();
```
Why not `contains()`?

cdumez commented on Source/WebCore/platform/graphics/avfoundation/objc/TextTrackPrivateMediaSourceAVFObjC.cpp:
```diff
@@ -0,0 +1,58 @@
+
+TextTrackPrivateMediaSourceAVFObjC::TextTrackPrivateMediaSourceAVFObjC(AVAssetTrack* track, CueFormat format)
+    : InbandTextTrackPrivateAVF(0, format)
+    , m_impl(makeUnique<AVTrackPrivateAVFObjCImpl>(track))
```
`makeUniqueRef()` and `UniqueRef` for the member type to make it clear it is never null?

philn commented on Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
```diff
@@ -117,7 +118,7 @@ MediaSourcePrivate::AddStatus MediaSourcePrivateRemote::addSourceBuffer(const Co
     // FIXME: Uses a new Connection for remote playback, and not the main GPUProcessConnection's one.
     // FIXME: m_mimeTypeCache is a main-thread only object.
-    callOnMainRunLoopAndWait([this, &returnedStatus, contentTypeString = contentType.raw().isolatedCopy(), &returnedSourceBuffer, gpuProcessConnection] {
+    callOnMainRunLoopAndWait([this, &returnedStatus, contentTypeString = contentType.raw().isolatedCopy(), &returnedSourceBuffer, gpuProcessConnection, configuration = WTFMove(configuration)] {
```
The configuration can't be moved, it's a const ref.

philn commented on Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
```diff
@@ -127,7 +128,7 @@ MediaSourcePrivate::AddStatus MediaSourcePrivateRemote::addSourceBuffer(const Co
         }
 
-        auto sendResult = gpuProcessConnection->connection().sendSync(Messages::RemoteMediaSourceProxy::AddSourceBuffer(WTFMove(contentType)), m_identifier);
+        auto sendResult = gpuProcessConnection->connection().sendSync(Messages::RemoteMediaSourceProxy::AddSourceBuffer(WTFMove(contentType), WTFMove(configuration)), m_identifier);
```
Ditto :)

eric-carlson committed:
[MSE][Mac] Add TextTrack support to SourceBufferPrivateAVFObjC.
https://bugs.webkit.org/show_bug.cgi?id=125161
rdar://137769578

Reviewed by Jean-Yves Avenard.

Add general TextTrack support to SourceBufferPrivateAVFObjC, and add support for muxed
WebVTT tracks.

* LayoutTests/media/media-source/content/bip-bop-webvtt-frag.mp4: Added.
* LayoutTests/media/media-source/media-managedmse-webvtt-track-expected.txt: Added.
* LayoutTests/media/media-source/media-managedmse-webvtt-track.html: Added.
* LayoutTests/platform/glib/TestExpectations: Skip new test.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Add new runtime setting.

* Source/WTF/wtf/Logger.h:
(WTF::LogArgument::toString): Support logging a StringView.

* Source/WebCore/Headers.cmake:

* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::createSourceBufferPrivate):

* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm: Add AVMediaTypeText.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/html/track/InbandTextTrack.cpp:
(WebCore::InbandTextTrack::create): Add `CueFormat::Unknown` to switch statement.

* Source/WebCore/html/track/InbandWebVTTTextTrack.cpp:
(WebCore::InbandWebVTTTextTrack::parseWebVTTFileHeader): Added.
* Source/WebCore/html/track/InbandWebVTTTextTrack.h:

* Source/WebCore/platform/graphics/InbandTextTrackPrivate.h: Add `CueFormat::Unknown`
so `m_format` can be initialized.

* Source/WebCore/platform/graphics/InbandTextTrackPrivate.serialization.in: Serialize
`CueFormat::Unknown`.

* Source/WebCore/platform/graphics/MediaSourceConfiguration.h:
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
* Source/WebCore/platform/graphics/TrackPrivateBase.cpp:
(WebCore::TrackPrivateBase::notifyMainThreadClient):

* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.h:
* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.mm:
(WebCore::AVTrackPrivateAVFObjCImpl::textKindForAVAssetTrack): Determine if a AVAssetTrack
has captions or subtitles based on its characteristics.
(WebCore::AVTrackPrivateAVFObjCImpl::textKindForAVMediaSelectionOption): Ditto, for an
AVMediaSelectionOption.
(WebCore::AVTrackPrivateAVFObjCImpl::textKind const): Return TextTrack.kind.

* Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
(WebCore::InbandTextTrackPrivateAVF::InbandTextTrackPrivateAVF): Make the AVFInbandTrackParent
parameter last and give it a default value of nullptr.
(WebCore::InbandTextTrackPrivateAVF::processCue): processNativeSamples -> processVTTSamples.
(WebCore::InbandTextTrackPrivateAVF::setMode): Always call through to the base class.
(WebCore::InbandTextTrackPrivateAVF::processVTTFileHeader): Split out of processNativeSamples.
(WebCore::InbandTextTrackPrivateAVF::processVTTSample): Ditto.
(WebCore::InbandTextTrackPrivateAVF::processVTTSamples): Renamed from processNativeSamples.
(WebCore::InbandTextTrackPrivateAVF::readVTTSampleBuffer): Renamed from readNativeSampleBuffer.
(WebCore::InbandTextTrackPrivateAVF::processNativeSamples): Deleted.
(WebCore::InbandTextTrackPrivateAVF::readNativeSampleBuffer): Deleted.
* Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.h:
(WebCore::InbandTextTrackPrivateAVF::setKind):
(WebCore::InbandTextTrackPrivateAVF::setId):
(WebCore::InbandTextTrackPrivateAVF::setLabel):
(WebCore::InbandTextTrackPrivateAVF::setLanguage):
(isType):

* Source/WebCore/platform/graphics/avfoundation/VideoTrackPrivateAVF.h: Remove an
unnecessary blank line.

* Source/WebCore/platform/graphics/avfoundation/objc/InbandTextTrackPrivateAVFObjC.h:
(WebCore::InbandTextTrackPrivateAVFObjC::create): Add version that takes an AVAssetTrack.
* Source/WebCore/platform/graphics/avfoundation/objc/InbandTextTrackPrivateAVFObjC.mm:
(WebCore::InbandTextTrackPrivateAVFObjC::InbandTextTrackPrivateAVFObjC): Switch parameter
order.
(WebCore::InbandTextTrackPrivateAVFObjC::disconnect): Call base class first, clear `m_assetTrack`.
(WebCore::InbandTextTrackPrivateAVFObjC::mediaType const):
(WebCore::InbandTextTrackPrivateAVFObjC::hasMediaCharacteristic const):
(WebCore::InbandTextTrackPrivateAVFObjC::kind const):
(WebCore::InbandTextTrackPrivateAVFObjC::isClosedCaptions const):
(WebCore::InbandTextTrackPrivateAVFObjC::isSDH const): Use hasMediaCharacteristic.
(WebCore::InbandTextTrackPrivateAVFObjC::containsOnlyForcedSubtitles const): Ditto.
(WebCore::InbandTextTrackPrivateAVFObjC::isMainProgramContent const): Ditto.
(WebCore::InbandTextTrackPrivateAVFObjC::isEasyToRead const): Ditto.
(WebCore::InbandTextTrackPrivateAVFObjC::label const): Simplify logic.
(WebCore::InbandTextTrackPrivateAVFObjC::language const): Use static methods in AVTrackPrivateAVFObjCImpl.
(WebCore::InbandTextTrackPrivateAVFObjC::isDefault const): Null-check m_mediaSelectionGroup
and m_mediaSelectionOption.

* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:
(WebCore::MediaSourcePrivateAVFObjC::addSourceBuffer):

* Source/WebCore/platform/graphics/avfoundation/objc/OutOfBandTextTrackPrivateAVF.h:
(WebCore::OutOfBandTextTrackPrivateAVF::OutOfBandTextTrackPrivateAVF):

* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferParserAVFObjC.mm:
(WebCore::MediaDescriptionAVFObjC::MediaDescriptionAVFObjC): Use -[AVAssetTrack mediaType]
instead of -[AVAssetTrack hasMediaCharacteristic] to get media type.
(WebCore::SourceBufferParserAVFObjC::SourceBufferParserAVFObjC): Add configuration parameter.
(WebCore::SourceBufferParserAVFObjC::didParseStreamDataAsAsset): Add support for WebVTT
tracks. Always log when we ignore an unknown track type.

* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::setTrackChangeCallbacks): Add a comment about why
we don't do anything when a text track mode changes.
(WebCore::SourceBufferPrivateAVFObjC::precheckInitializationSegment): Support text tracks.
(WebCore::SourceBufferPrivateAVFObjC::isMediaSampleAllowed const): Deal with text samples.
(WebCore::SourceBufferPrivateAVFObjC::clearTracks): Support text tracks.
(WebCore::SourceBufferPrivateAVFObjC::trackDidChangeEnabled): Fix indentation.
(WebCore::SourceBufferPrivateAVFObjC::isTextTrack const):

* Source/WebCore/platform/graphics/avfoundation/objc/TextTrackPrivateMediaSourceAVFObjC.cpp: Added.
(WebCore::TextTrackPrivateMediaSourceAVFObjC::TextTrackPrivateMediaSourceAVFObjC):
(WebCore::TextTrackPrivateMediaSourceAVFObjC::resetPropertiesFromTrack):
* Source/WebCore/platform/graphics/avfoundation/objc/TextTrackPrivateMediaSourceAVFObjC.h:

* Source/WebCore/platform/graphics/cocoa/SourceBufferParser.cpp:
(WebCore::SourceBufferParser::create): Add configuration parameter.
* Source/WebCore/platform/graphics/cocoa/SourceBufferParser.h:

* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.cpp:
(WebCore::MediaSourcePrivateGStreamer::addSourceBuffer): Ditto
* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.h:

* Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.cpp:
(WebCore::MockMediaSourcePrivate::addSourceBuffer): Ditto.
* Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.h:

* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::loadMediaSource): Remove webMParserEnabled parameter.
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:

* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:
(WebKit::RemoteMediaSourceProxy::RemoteMediaSourceProxy):
(WebKit::RemoteMediaSourceProxy::addSourceBuffer): Add configuration parameter.
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.messages.in:

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in: Add MediaSourceConfiguration.

* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::load): Add configuration parameter.

* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
(WebKit::MediaSourcePrivateRemote::addSourceBuffer): Ditto.
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.h:

Canonical link: https://commits.webkit.org/290705@main
